### ISSUE TITLE
Is it possible to give stage or environment based access?

### ISSUE THREAD

**Issue by danihodovic** [zendesk/samson#2840]
I would like to restrict deployment access per stage or environment. An example use case is to allow junior developers to deploy to staging and senior developers to deploy to production.

Is this possible? I've looked around in the docs and played with the UI and I found nothing.

**Comment by grosser** [comment]
not possible, sounds like an interesting feature, but might take a lot of new logic/UI to add :/

/cc @zendesk/bre

**Comment by jason-rutherford** [comment]
One possible workaround might be to lock the stage and use the `PRODUCTION_STAGE_LOCK_REQUIRES_ADMIN=1` feature [here](https://github.com/zendesk/samson/pull/2492/files/fd5f7b393c20ab789008b9eeecbad78e99fcdea1) . I'm not sure if that helps in your situation but I thought it is worth mentioning.

**Comment by danihodovic** [comment]
@jason-rutherford That would require us to lock production on every deploy which is not ideal.

**Comment by grosser** [comment]
Possible solutions:
 - add an option to only allow admins to deploy to production (effort: low / hackyness: medium)
 - add a new 'production-deployer' role between deployer and admin (effort: medium / hackyness: medium)
 - overhaul the permission system to be finer-grained / allow random things like "deploy production" / "manage secrets" etc  (effort: high / hackyness: low)

**Comment by danihodovic** [comment]
> overhaul the permission system to be finer-grained / allow random things like "deploy production" / "manage secrets" etc (effort: high / hackyness: low)

Sounds like the best option, but if we're the only team with this use-case it might be better to leave it as is and not introduce additional complexity.

**Comment by grosser** [comment]
yeah ... we've come a long way with this simple system and reworking it to
accommodate 1 new usecase seems a bit extreme

**Comment by ibrahima** [comment]
Hmm, I'd be interested in this for the same reason. The current state is fine as long as you trust your teammates to not do stuff they shouldn't be doing, but it'd make me feel a little safer if junior developers couldn't deploy to production (this is more of an issue when adding new teammates IMO).

**Comment by grosser** [comment]
can enable BUDDY_CHECK_FEATURE=1 so nobody can deploy to prod alone :)

**Comment by ibrahima** [comment]
Hmm, does that apply to admins too? We aren't using buddy checks but that could be useful for us.

**Comment by grosser** [comment]
applies to everyone ... since it's kinda good practice ... can be bypassed for emergencies, then need to setup email for notification
